3단계 - 로또(자동) 과제 PR 요청드립니다!

taewonh opened the pull request:
늦은 시간에 PR 드려 죄송합니다.
이번 과제를 진행하면서 지난 프리코스 과정보다 훨씬 더 복잡한 과제라는것을 느낄 수 있었는데요.. 특히 로또 번호에 대한 검증 결과를 담는 클래스 설계를 어떤식으로 진행할지 고민을 많이 했는데 좋은 결과인지는 확신이 안서네요 .. ㅜ
(사실 4번째 과제에 적용할 수 있는 Rank Enum 을 이번 과제에 적용하면 조금 더 깔끔한 구조가 나올 수 있지 않을까해서 고민을 많이 했었는데 다음 과제를 위해 적용하지 않았습니다..)

과제를 진행하면서 조금 아쉬웠던 점은 초반에는 TDD를 지키다가 구조상 마음에 안드는 부분도 생기고 코드 양도 많아지면서 점점 테스트 코드보다 프로덕션 코드를 우선해서 뭔가 자연스럽게 손을 댄 부분이 있었는데.. 그래도 최대한 테스트 코드를 먼저 건들려고는 시도했습니다..

또 진행하면서 한가지 궁금한점이 있는데 Controller와 InputView, OutputView 를 작성면서 전체 기능 실행의 주체가 되는 것이 Controller고 이러한 Controller에서 InputView와 OutputView를 필연적으로 사용하게 되는 것 같은데, InputView, OutputView를 사용하는 주체가 Controller가 되는것이 맞는 건지, 이게 좋은 방식인지 아닌지가 궁금합니다!

taewonh committed:
chore: 이전 진행 과제 패키지 분리

- 이후부터 진행하는 과제 분리를 위해 이제까지 진행했던 과제에 대해 단계별로 패키지 분리

taewonh committed:
docs(README.md): step3 핵심 기능 목록 작성

taewonh committed:
feat: 로또 번호 자동 생성 기능 구현

- LottoNumbers 클래스 추가
  - 정수 리스트를 가지는 일급 컬렉션
  - new 키워드 사용 금지
  - generate 메소드 파라미터로 시작 번호, 끝 번호, 번호 선택 개수를 입력하여 자동 번호 생성

- Lotto 클래스 추가
  - LottoNumbers와 번호 자동 생성을 위한 필요 정보(시작 번호, 끝 번호, 번호 선택 개수)를 멤버로 가짐
  - new 키워드 사용 금지

taewonh committed:
refactor: 로또 번호 자동 생성 주체 변경

- 번호를 담는 List 생성은 LottoNumbers가 아닌 Lotto 에서 생성하도록 담당하고, 생성된 List를 주입하는 방식으로 변경
  - LottoNumbers는 단순히 List만 멤버로 가지고 해당 List에 대한 검증이나 비교 처리에 대한 역할만 가지는게 적합하다고 판단
- 로또 번호 생성 테스트 시 값 비교를 위해 Lotto의 상수를 모두 public 으로 처리
  - 변경이 불가능한 값이기 때문에 외부에서 조회하는 것은 특별히 문제될 것이 없어 보임
- 로또 번호 자동 생성 테스트 시 생성한 번호의 개수를 비교하는 테스트 로직 추가

taewonh committed:
docs(README.me): 로또 번호 자동 생성 내용 수정

- 로또 번호 자동 생성 주체 클래스 변경에 따른 기능 내용 수정

taewonh committed:
refactor(Lotto): 로도 번호 자동 생성 로직 stream 적용

- for문이 아닌 IntStream 적용 및 리스트 변환시 collectingAndThen을 적용하여 shuffle, subList, sort 처리 적용

taewonh committed:
refactor(LottoNumbers): 중복 숫자 목록 입력 예외 처리

- 중복 숫자 목록 입력시 IllegalArgumentException throw 처리

taewonh committed:
feat: 당첨 번호 비교 기능 구현

- LottoNumbers 간 match 메소드를 이용하여 일치하는 번호 개수를 추출하는 기능 추가
- 번호 비교시 각 비교대상의 번호가 다른 경우 예외 처리

taewonh committed:
feat: 로또 구입 기능 구현

- 입력받은 총 금액을 기준으로 구입 가능한 최대 개수 구입
- 로또 구입 후 Lotto List를 저장하는 1급 컬렉션(Lottos) 반환
- 구입 금액이 부족한 경우 예외처리

taewonh committed:
refactor: 전체 코드 일괄 리팩토링 진행

- 도메인 역할을 하는 클래스들을 domain 패키지로 분리
- Lotto 랜덤 번호 생성을 수행하는 별도의 Generator 추가
- Lotto 인스턴스 생성시 외부에서 Numbers를 주입하도록 설정
- LottoNumbers -> Numbers 클래스명 변경

taewonh committed:
feat: 당첨 결과 생성 기능 구현

- 구입한 로또 객체와 당첨 번호를 비교하여 당첨 결과(Reward) 생성
- 당첨 정보 생성에 필요한 기준 정보는 당첨 결과에서 보유 (추후 리팩토링을 통해 외부 주입 방식으로 변경 예정)
- 당첨 기준 정보 참고후 당첨이 되지 않은 경우 객체를 생성하지 않고 Exception Throw 처리

taewonh committed:
refactor: 당첨 결과 생성 기능 구조 리팩토링

- 기존 생각하던 방식인 Reward, Reward 보다 적합해 보이는 방식으로 변경
- 구입한 로또 목록과 당첨 번호, 당첨에 사용되는 기준 정보를 바탕으로 결과 저장
      - 당청 결과는 Mpa<Integer, Integer> 형식으로 번호 일치 개수와 전체 수의 쌍으로 저장
    - 당첨에 사용되는 기준 정보는 CriteriaProvider라는 함수형 인터페이스를 사용하여 외부에서 주입하여 사용
    - 각 matchCount별로 문자형 형식의 통계 정보 생성

taewonh committed:
chore(Numbers): 기본 생성자 사용시 exception throw 처리

- 외부에서 사용이 불가능하긴 하지만 명시적으로 사용 금지라는 것을 알려주기 위해 빈 ArrayList로 지정하는 것이 아닌 Exception을 Throw하는 것으로 변경

taewonh committed:
refactor(LottoStore): 구매 가능 금액 확인 로직 별도 메소드 분리

taewonh committed:
refactor(Reward): 전체 리팩토링

- Reward 내부의 결과를 저장하는 Map 명칭을 reward로 변경
- Reward 정보 생성시 Lottos와 winningNumbers로 생성하는 것이 아닌 matchCounts만 사용하도록 개선
   - 기능 쪼개기 및 테스트 코드를 조금 더 명확하게 작성하기 위해
- 테스트 코드 전체 개선
  - 쉽게 예측 가능한 값들을 직접 사용하도록 변경

taewonh committed:
feat(Reward): 수익률 계산 기능 구현

- 전체 구매 금액을 입력 받아 총 당첨 금액과 계산하여 수익률 계산을 구현
- 총 당첨 금액 계산 기능은 외부에서 사용할 수 없어 private 처리 및 테스트 코드 또한 수익률 계산 기능 테스트 코드로 변경
- 그외 해당 기능 관련 명칭에 대해 리팩토링 진행 (의미상 조금 더 부합하는 이름으로 변경)

taewonh committed:
feat: 로또 구매 및 결과 출력 전체 기능 구현

- Application을 실행하는 메인 클래스 추가 (Application)
- 문자열 입출력을 담당하는 InputView, OutputView 추가
- Lotto 전체 기능을 실행을 통제하는 LottoController 추가
  - LottoController 생성 시 InputView, OutputView, LottoStore, CriteriaProvider(로또 당첨 확인에 사용되는 기준 정보)
  - 각 기능(로또 구입 금액 지불, 로또 구매, 당첨 번호 입력, 결과) 총 4가지 메소드로 분리하여 전체 기능 실행
- 금액으로 사용되는 long 타입을 가지는 WrapperClass 추가
  - Money 클래스 추가 및 해당 long 타입을 사용하는 곳 모두 일괄 변경 적용
- 일부 메소드, 멤버 명칭을 의미상 조금더 명확하게 개선

heowc commented on src/main/java/step3/domain/Money.java:
```diff
@@ -0,0 +1,31 @@
+        return new Money(money);
+    }
+
+    public void validate(Money pricePerLotto) {
```
입력한 금액으로 로또를 1개 이상 구매할 수 있는지를 검증하는 메소드로 보이는데요.
여기서 로또 금액 알 필요가 있을까요? 그로 인해 구매할 수 있는지에 대한 검증이 포함된 것이라고 판단됩니다.

`Money`가 가져야할 책임을 무엇일까요? 🤔

taewonh replied:
Money는 단순히 나눌값(로또 1장당 구매 가격)을 외부(LottoStore)에서 입력받아 나누는 기능만 구현하고, 해당 값을 사용하는 외부(LottoStore) 에서 validate 하는 방식으로 변경했습니다!

heowc commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,27 @@
+
+    private Lotto(Numbers numbers) {
+        this.numbers = numbers;
+    }
```
기본 생성자를 만들어 예외처리를 하신 이유가 있을까요? 🤔 
별도의 생성자를 만들었다면 기본 생성자를 굳이 만들어 예외처리를 할 필요가 없어보여서요.

taewonh replied:
음.. 우선 디폴트 생성자를 예외처리한 클래스가 Lotto, Numbers, Reward 클래스가 있는데, 이렇게 한 이유는 의존성이 필요한 클래스의 인스턴스를 만들때 반드시 생성 시점에 주입해서 사용하게 강제화 시키고자 함입니다..

setter가 없기 때문에 만약 디폴트 생성자를 public으로 풀더라도 의존성 주입을 할 방법이 없어 어차피 사용 불가능한 인스턴스일텐데 이런 부분까지는 굳이 생각할 필요가 없을까요...?
(막상 이렇게 생각해놓고 Lottos는 누락시켰네요.. 하하..)

heowc replied:
자바 스펙상 임의의 생성자를 추가했다면 디폴트 생성자를 사용할 수 없습니다. 그럼에도 디폴트 생성자를 작성한 이유가 있나해서요 🤔

taewonh replied:
앗 이부분은 기본기 부족인거 같네요... 반성하겠습니다 ..

heowc replied:
서로서로 배워가는거죠 😄 
저도 리뷰를 통해 많이 배우고 있습니다 👍

heowc commented on src/main/java/step3/domain/Numbers.java:
```diff
@@ -0,0 +1,46 @@
+        return (int) numbers.stream()
+                .filter(selectNumbers.numbers::contains)
+                .count();
+    }
```
1. 중복되지 않는 숫자를 갖는지 검증
2. `Numbers` 포함된 숫자 갯수가 동일하지 검증

해당 메소드는 `Lotto`의 기능이지 않을까요? 어떻게 생각하시나요 🤔

taewonh replied:
위의 로또 금액 검증 피드백과 같은 맥락으로 이해하고 아래와 같이 처리하였습니다.

1. Numbers는 가진 값중 중복되었는지 여부만 확인하는 기능만 제공하고 Lotto 에서 해당 기능을 이용하여 validate 처리
2. Numbers는 외부에서 다른 Numbers를 파라미터로 입력받아 사이즈가 동일한지만 확인해주고, Lotto에서 해당 기능을 이용하여 validate 처리

확실히 각 클래스가 가져야 할 책임이 조금더 명확해진 것 같긴하네요!

heowc replied:
`Numbers`에서 중복 체크를 하고자 한다면, `UniqueNumbers`라는 클래스로 만드는게 조금 더 명확하지 않을까요? 🤔

heowc commented on src/main/java/step3/domain/Reward.java:
```diff
@@ -0,0 +1,74 @@
+                .collect(Collectors.toList());
+    }
+
+    private String generateStatistic(int matchCount) {
```
Reward에서는 결과만 주고, 그걸 어떻게 표현할지는 View에 맞기는 건 어떨까요? 🤔

taewonh replied:
말씀하신대로 Reward는 순수한 데이터만 외부로 제공하고 OutputView에서 해당 값들을 조합하여 문자열을 만들고 print 하는 방식으로 처리 완료했습니다!

heowc commented on src/main/java/step3/utils/LottoNumbersGenerator.java:
```diff
@@ -0,0 +1,30 @@
+    public static final int END_NUMBER = 45;
+    public static final int MAX_SELECT_NUMBER = 6;
+
+    public static Numbers random() {
```
하나의 메소드에서는 10 라인을 넘지 않도록 하는 것이 요구사항에 있습니다. 🙇 
이를 어떻게 줄일 수 있을까요?

taewonh replied:
스트림 내부에서 처리되는 로직들도 라인 제약사항으로 들어가나 보군요.. 

우선 45개 번호들만을 담고있는 리스트를 제공하는 메소드를 추가로 만들어 분리하여 라인수 초과되는 부분을 제거 처리했습니다!

heowc commented on src/test/java/step1/SetTest.java:
```diff
@@ -1,3 +1,5 @@
+package step1;
```
적절한 패키지 분리 좋습니다 👍

heowc commented on src/main/java/step3/Application.java:
```diff
@@ -0,0 +1,30 @@
+        InputView inputView = new InputView();
+        OutputView outputView = new OutputView();
+        LottoStore lottoStore = new LottoStore();
+        CriteriaProvider provider = () -> {
```
> 사실 4번째 과제에 적용할 수 있는 Rank Enum 을 이번 과제에 적용하면 조금 더 깔끔한 구조가 나올 수 있지 않을까해서 고민을 많이 했었는데 다음 과제를 위해 적용하지 않았습니다.

enum 대신에 `CriteriaProvider`라는 클래스를 만들어 사용하셨군요 👍 
정답은 없지만 `enum`을 사용하셔도 되고 그 외 방법으로도 사용해도 무방합니다. 

다만, 지금의 `CriteriaProvider`이 반환하는 `Map`은 key/value가 어떤걸 값을 의미하는지 판단하기 어려운 것 같아요.

taewonh replied:
넵넵 말씀하신 대로 의미가 불분명해보여 확실히 가독성도 떨어지고 코드도 쓸떼없이 지저분해지는 느낌이더라구요..

일단 Rank Enum을 적용해서 전체적으로 다 변경하고 CriteriaProvider는 제거 처리하였습니다!

heowc commented on src/main/java/step3/domain/Lottos.java:
```diff
@@ -0,0 +1,31 @@
+                .collect(Collectors.toList());
+    }
+
+    public String toString() {
```
```suggestion
    @Override
    public String toString() {
```

taewonh replied:
앗.. 누락이 있었군요..

heowc commented on src/main/java/step3/domain/LottoStore.java:
```diff
@@ -0,0 +1,20 @@
+
+    public static final Money pricePerLotto = Money.generate(1000);
+
+    public Lottos purchase(Money payment) {
```
`로또 상점`에 `구매한다` 라는 메소드가 있는 점이 다소 어색하다고 생각합니다. 어떻게 생각하시나요? 🤔 

판매한다/교환한다 등은 어떨까요?

taewonh replied:
개발하다보니 자연스럽게 제가 로또를 구입해서 결과를 보는 사람으로 생각하다보니 저렇게 된거 같습니다 ... 항상 객체를 설계할 때는 두번 세번 꼼꼼하게 생각해야될것 같네요! 감사합니다!

heowc reviewed (changes requested):
안녕하세요. 태원님 

늦은 시간까지 3단계를 진행해주셨네요! 고생 많으셨습니다. 🙇 

전체적으로 잘 진행해주셨지만, 조금 더 고민해보면 좋을 같은 내용이 있어 코멘트 남겼습니다 🙏 

코멘트 보시고 궁금하신 점이나 함께 이야기해보면 좋을 내용이 있다면 깃헙 코멘트 또는 슬랙 DM으로 편하게 말씀 부탁드립니다 🙇

---- 

> 과제를 진행하면서 조금 아쉬웠던 점은 초반에는 TDD를 지키다가 구조상 마음에 안드는 부분도 생기고 코드 양도 많아지면서 점점 테스트 코드보다 프로덕션 코드를 우선해서 뭔가 자연스럽게 손을 댄 부분이 있었는데.. 그래도 최대한 테스트 코드를 먼저 건들려고는 시도했습니다..

고민해시고 노력하는 모습 보기 좋은 것 같습니다 👍 
처음에는 어려울 수 있습니다. 경험에 따라 다르겠지만, 보통은(?) 도메인 중심의 개발을 하지 않는 것 경험이 더 많이 때문에 이번 기회로 `객체의 책임/관계`에 대해서 고민해보신다면 보다 성장하실 수 있을 거에요. 화이팅입니다! 👍  

> 또 진행하면서 한가지 궁금한점이 있는데 Controller와 InputView, OutputView 를 작성면서 전체 기능 실행의 주체가 되는 것이 Controller고 이러한 Controller에서 InputView와 OutputView를 필연적으로 사용하게 되는 것 같은데, InputView, OutputView를 사용하는 주체가 Controller가 되는것이 맞는 건지, 이게 좋은 방식인지 아닌지가 궁금합니다!

View의 역할을 충분히 잘 분리해주셨고(비즈니스 로직이 포함되지 않은..?) 서로간의 참조되지 않는 구조이기 때문에 좋다고 생각합니다. 👍 

말씀해신 `Controller` 와 `{Input/Output}View`는 MVC 패턴을 고민해보시면 정답이 나올 거라고 생각합니다.
https://developer.mozilla.org/ko/docs/Glossary/MVC

taewonh committed:
refactor(LottoStore): 로또 판매 메소드 명칭 변경

- LottoStore 기준으로 로또를 판매하는 것이기 때문에 purchase(구매하다)가 아닌 sell(판매하다)로 변경

taewonh committed:
refactor: Controller의 누락된 로또 구매 메서드 변경점 반영

taewonh committed:
refactor: 번호 생성기 클래스 명칭 변경

- LottoNumbersGenerator가 아닌 NumbersGenerator로 변경

taewonh committed:
chore: Lottos.toString 메소드의 override 어노테이션 추가

taewonh committed:
fix: 번호 생성 버그 수정

- 리팩토링 과정에서 발생한 오류로써, 선택한 번호 개수만큼이 아닌 최대(45) 개수 만큼 생성되는 문제 수정

taewonh committed:
refactor: 로또 구매가 가능한지에 대한 검증 부분 변경

- 기존 Money에서 로또 1장당 금액을 입력받아 검증하는 부분을 LottoStore에서 담당하도록 변경
- Money는 단순히 전체금액 / 1장당 금액에 대한 결과만을 return하는 divide 기능만을 가지도록 기능 변경

taewonh committed:
refactor: 중복 로또 번호 검증 주체 변경

- Numbers가 아닌 Lotto 객체에서 중복 로또 번호를 검증하도록 변경
- Numbers는 단순히 중복된 번호가 존재하는지 여부를 return 하는 기능만 제공하고 Lotto가 해당 기능을 이용하여 검증을 수행함

taewonh committed:
refactor: 일치하는 로또 번호 개수 추출시 검증 주체 변경

- 일치하는 로또 번수 개수 추출시 검증에 대한 주체를 Numbers가 아닌 Lotto로 변경
- Numbers는 단순히 size가 동일한지에 대한 비교만 수행

taewonh committed:
chore: 제거가 누락된 import 제거

taewonh committed:
refactor: 로또 당첨 개수 및 금액 비교 정보 클래스 변경

- 기존 사용하던 Map<Integer, Integer>이 아닌 Enum을 사용하는 것으로 변경
- 더이상 Map이 사용되지 않아 CriteriaProvider 제거 및 주입 파라미터 모두 제거

taewonh committed:
refactor: Lottos 내부에서 List<Integer>가 아닌 List<Rank>를 사용하도록 변경

- 추가적으로 countMatch 변수명을 Rank 내부에서 사용하는 matchOfCount와 동일하게 변경
- Lottos.getRank 시 당첨 금액이 있는 경우에 한하여 List에 입력하도록 변경 (메모리 과다 소모 해결)

taewonh committed:
refactor: 당첨 통계 정보 생성 및 확인 방식 변경

- 기존 Reward에서 당첨 통계 정보를 이용해서 문자열을 생성하는 것이 아닌 Reward는 당첨 결과 데이터만을 외부로 전달하고 OutputView에서 해당 정보들을 조합하여 print하는 방식으로 변경

taewonh committed:
refactor: 수익률 print 파라미터 변경

- 가독성문제로 인해 파라미터로 수익률을 계산해서 전달하는 것이 아닌 Money와 Reward 만을 전달하고 메소드 내부에서 계산하도록 변경

heowc commented on src/main/java/step3/domain/Lotto.java:
```diff
@@ -0,0 +1,41 @@
+    public String toString() {
+        return "[" + numbers.toString() + "]";
+    }
+}
```
EOL 을 챙겨주시면 좋을 것 같습니다 👍 
https://velog.io/@d-h-k/intellij-%ED%8C%8C%EC%9D%BC%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-%EC%B6%94%EA%B0%80%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95

heowc commented on src/main/java/step3/domain/LottoStore.java:
```diff
@@ -0,0 +1,26 @@
+    public Lottos sell(Money payment) {
+        validate(payment);
+        List<Lotto> lottoList = new ArrayList<>();
+        for (int count = 0; count < payment.divide(pricePerLotto); count++) {
```
`payment.divide(pricePerLotto)`를 지역변수로 바꾸는게 좋지 않을까요? 🤔

heowc commented on src/main/java/step3/domain/Lottos.java:
```diff
@@ -0,0 +1,34 @@
+
+public class Lottos {
+
+    private List<Lotto> lottoList;
```
특정 구현을 지칭하는 변수명은 지향하는 것이 좋다고 생각합니다.
복수를 나타내는 `lottos`라고 하는건 어떨까요? 🤔

heowc commented on src/main/java/step3/domain/Rank.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public static Rank valueOf(int countOfMatch) {
+        return Arrays.stream(Rank.values())
+                .filter(rank -> rank.countOfMatch == countOfMatch)
+                .findFirst()
+                .orElseGet(() -> Rank.MISS);
+        // TODO 일치하는 수를 로또 등수로 변경한다. enum 값 목록은 "Rank[] ranks = values();"와 같이 가져올 수 있다.
```
1. 불필요한 주석은 제거하는게 좋을 것 같습니다. 👍 
2. `Optional#orElseGet` 과 `Optional#orElse`의 차이를 아시나요?

taewonh replied:
음 말씀해주신대로 차이를 한번 찾아봤는데 저 코드의 의도가 countOfMatch와 해당되는 Rank가 없는 경우에만 Rank.MISS 를 외부로 return 하고자 함이기 때문에 orElseGet을 사용하는 것이 맞는 것 같습니다.

만약 orElse로 지정이 되었다하더라도 Rank.MISS를 주기 때문에 동작에는 문제가 없었겠지만, 해당하는 Rank가 있을때도 orElse 내부의 구문이 불필요하게 호출이 되기 때문에 저 상황에서는 orElseGet을 사용하는 것이 맞다고 생각됩니다.

특히 실무에서 orElse 내부에서 불필요하게 DB 작업같은 것들을 실행하게 되면 문제가 생길 소지가 있어보여 역시 제대로 알고 사용하는 것이 중요하겠네요!

heowc replied:
👍👍

heowc commented on src/main/java/step3/domain/Reward.java:
```diff
@@ -0,0 +1,67 @@
+        return reward.keySet()
+                .stream()
+                .map(this::getWinningMoney)
+                .mapToLong(money -> money)
```
해당 라인은 제거해도 되지 않을까요? 🤔

heowc commented on src/test/java/step3/domain/LottoStoreTest.java:
```diff
@@ -0,0 +1,31 @@
+        Money payment = Money.generate(0);
+        assertThatThrownBy(() -> {
+            Lottos lottos = store.sell(payment);
+        })
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("You don't have enough money.");
```
```suggestion
        })
        .isInstanceOf(IllegalArgumentException.class)
        .hasMessageContaining("You don't have enough money.");
```

heowc commented on src/test/java/step3/domain/NumbersTest.java:
```diff
@@ -0,0 +1,39 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@TestInstance(Lifecycle.PER_CLASS)
```
해당 애노테이션은 어떤 이유로 사용하신 걸까요 🤔

heowc reviewed (approved):
리뷰 반영 감사합니다 🙇

이번 단계는 여기서 마무리 하도록 할게요
이전보다 코드 읽기가 좋아진 것 같습니다 👍👍  

약간의 소소한(?) 코멘트를 남겨봤는데요. 다음 단계 진행시 반영 및 코멘트 부탁드릴게요 🙏
코멘트 보시고 궁금하신 점이나 함께 이야기해보면 좋을 내용이 있다면 깃헙 코멘트 또는 슬랙 DM으로 편하게 말씀 부탁드립니다 🙇
